In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os
import re

In [2]:
os.getcwd()

'C:\\Users\\sunilchawla\\Desktop\\ass2\\code'

In [3]:
path = "../data_df/"

In [4]:
df_train = pd.read_csv(path+"pos_neg.csv",sep = ",")
df_test = pd.read_csv(path+"pos_neg_test.csv",sep = ",")

In [5]:
df_train.drop(['Unnamed: 0'], axis = 1, inplace = True)
df_test.drop(['Unnamed: 0'], axis = 1, inplace = True)

In [6]:
df_train.head()

movieID                                             review  rate  label
0        0  Story of a man who has unnatural feelings for ...     3      0
1    10000  Airport '77 starts as a brand new luxury 747 p...     4      0
2    10001  This film lacked something I couldn't put my f...     4      0
3    10002  Sorry everyone,,, I know this is supposed to b...     1      0
4    10003  When I was little my parents took me along to ...     1      0

In [7]:
df_test.head()

movieID                                             review  rate  label
0        0  Once again Mr. Costner has dragged out a movie...     2      0
1    10000  This is an example of why the majority of acti...     4      0
2    10001  First of all I hate those moronic rappers, who...     1      0
3    10002  Not even the Beatles could write songs everyon...     3      0
4    10003  Brass pictures (movies is not a fitting word f...     3      0

In [8]:
X_train=df_train['review'].str.lower()
X_test=df_test['review'].str.lower()
Y_train = df_train['label']
Y_test = df_test['label']

# <font color="red">Feature Extraction using countVectorizer</font>

In [9]:
from sklearn.feature_extraction.text import CountVectorizer
count_vect = CountVectorizer(binary = True)
train_document = count_vect.fit_transform(X_train)
test_document = count_vect.transform(X_test)

In [89]:
train_document

<25000x74849 sparse matrix of type '<class 'numpy.int64'>'
	with 3445861 stored elements in Compressed Sparse Row format>

# <font color="red">DecisionTreeClassifier Sentiment Analysis(on countVectorizer feature)</font>

In [13]:
from sklearn.tree import DecisionTreeClassifier
clf = DecisionTreeClassifier().fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.2f}".format(clf.score(test_document, Y_test)))

Training set score: 1.00
Test set score: 0.71


# <font color="red">RandomForestClassifier Sentiment Analysis(on countVectorizer feature)</font>

In [12]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier().fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.2f}".format(clf.score(test_document, Y_test)))

Training set score: 0.99
Test set score: 0.74


# <font color="red">Naive Bayes Sentiment Analysis(on countVectorizer feature)</font>

In [90]:
from sklearn.naive_bayes import BernoulliNB
classifier = BernoulliNB().fit(train_document,Y_train)

In [91]:
print("Training set score: {:.2f}".format(classifier.score(train_document, Y_train)))
print("Test set score: {:.2f}".format(classifier.score(test_document, Y_test)))

Training set score: 0.90
Test set score: 0.83


# <font color="red">Support vector machine(SVM) Sentiment Analysis (on countVectorizer feature)</font>

In [44]:
from sklearn.svm import LinearSVC

In [93]:
clf = LinearSVC().fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.2f}".format(clf.score(test_document, Y_test)))

Training set score: 1.00
Test set score: 0.85


# <font color="red">Logistic Regression Sentiment Analysis(on countVectorizer feature)</font>

In [94]:
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression().fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.2f}".format(clf.score(test_document, Y_test)))

Training set score: 1.00
Test set score: 0.87


# <font color="red">SGD Sentiment Analysis(on countVectorizer feature)</font>

In [100]:
from sklearn.linear_model import SGDClassifier
clf = SGDClassifier().fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.2f}".format(clf.score(test_document, Y_test)))

C:\Users\sunilchawla\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


Training set score: 0.94
Test set score: 0.84


# <font color="red">Gradientboosting Sentiment Analysis(on countVectorizer feature)</font>

In [11]:
from sklearn.ensemble import GradientBoostingClassifier
clf = GradientBoostingClassifier().fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.2f}".format(clf.score(test_document, Y_test)))

Training set score: 0.82
Test set score: 0.81


# <font color="red">Feature Extraction using TF-IDF</font>

In [14]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf_vect = TfidfVectorizer()
train_document = tfidf_vect.fit_transform(X_train)
test_document = tfidf_vect.transform(X_test)

# <font color="red">DecisionTreeClassifier Sentiment Analysis (on TF-IDF feature)</font>

In [15]:
from sklearn.tree import DecisionTreeClassifier
clf = DecisionTreeClassifier().fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.2f}".format(clf.score(test_document, Y_test)))

Training set score: 1.00
Test set score: 0.70


# <font color="red">RandomForestClassifier Sentiment Analysis (on TF-IDF feature)</font>

In [16]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier().fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.2f}".format(clf.score(test_document, Y_test)))

Training set score: 0.99
Test set score: 0.73


# <font color="red">Gradient Boosting Classifier Sentiment Analysis (on TF-IDF feature)</font>

In [17]:
from sklearn.ensemble import GradientBoostingClassifier
clf = GradientBoostingClassifier().fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.2f}".format(clf.score(test_document, Y_test)))

Training set score: 0.83
Test set score: 0.81


# <font color="red">Support vector machine(SVM) Sentiment Analysis (on TF-IDF feature)</font>

In [96]:
clf = LinearSVC().fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.2f}".format(clf.score(test_document, Y_test)))

Training set score: 0.99
Test set score: 0.88


# <font color="red">Logistic Regression Sentiment Analysis(on TF-IDF feature)</font>

In [97]:
clf = LogisticRegression().fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.2f}".format(clf.score(test_document, Y_test)))

Training set score: 0.93
Test set score: 0.88


# <font color="red">SGD Sentiment Analysis (on TF-IDF feature)</font>

In [111]:
from sklearn.linear_model import SGDClassifier
clf = SGDClassifier().fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.2f}".format(clf.score(test_document, Y_test)))

Training set score: 0.94
Test set score: 0.89


C:\Users\sunilchawla\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


# <font color="red">Feature Extraction using Hashing Vectorizer</font>

In [105]:
from sklearn.feature_extraction.text import HashingVectorizer
hash_vect = HashingVectorizer()
train_document = hash_vect.fit_transform(X_train)
test_document = hash_vect.transform(X_test)

#   

# hyperparameter tuning

In [11]:
from sklearn.model_selection import GridSearchCV

# <font color="red">hyperparameter tuning in multinomial naive bayes algorithm</font>

In [14]:
from sklearn.naive_bayes import MultinomialNB
param_grid = {'alpha': [0.01, 0.1, 1, 10]}
grid_search = GridSearchCV(MultinomialNB(),param_grid, cv = 3)
grid_search.fit(train_document, Y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'alpha': [0.01, 0.1, 1, 10]}, pre_dispatch='2*n_jobs',
       refit=True, return_train_score='warn', scoring=None, verbose=0)

### <font color="green">large alpha => more smoothing => result in less complex model</font>

In [15]:
clf = MultinomialNB(alpha=1.0, fit_prior=True, class_prior=None).fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.2f}".format(clf.score(test_document, Y_test)))

Training set score: 0.91
Test set score: 0.83


# <font color="red">hyperparameter tuning in SVM algorithm</font>

<br><font color="green">gamma =>control the width of kernel</font> </br>
<br><font color="green">C =>Regularization => limits the importance of each word</font></br>
<br><font color="green">gamma and C both control the complexity of the model, with large values</font></br>
<br><font color="green"> in either resulting in a more complex model</font></br>

In [31]:
param_grid = {'C': [1, 10],'gamma': [0.001, 0.01, 0.1, 1],'kernel': ['linear']}

In [32]:
from sklearn.svm import SVC
grid_search = GridSearchCV(SVC(),param_grid, cv = 5)
grid_search.fit(train_document, Y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'C': [1, 10], 'gamma': [0.001, 0.01, 0.1, 1], 'kernel': ['linear']},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [45]:
clf = LinearSVC(C=1.0).fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.2f}".format(clf.score(test_document, Y_test)))

Training set score: 0.99
Test set score: 0.88


# <font color="red">hyperparameter tuning in SGD algorithm</font>

In [27]:
param_grid = {'alpha': [0.00001,0.0001, 0.001, 0.01, 0.1, 1, 10]}
from sklearn.linear_model import SGDClassifier
grid_search = GridSearchCV(SGDClassifier(),param_grid, cv = 3)
grid_search.fit(train_document, Y_train)

C:\Users\sunilchawla\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)
C:\Users\sunilchawla\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." %

C:\Users\sunilchawla\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)
C:\Users\sunilchawla\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." %

GridSearchCV(cv=3, error_score='raise',
       estimator=SGDClassifier(alpha=0.0001, average=False, class_weight=None, epsilon=0.1,
       eta0=0.0, fit_intercept=True, l1_ratio=0.15,
       learning_rate='optimal', loss='hinge', max_iter=None, n_iter=None,
       n_jobs=1, penalty='l2', power_t=0.5, random_state=None,
       shuffle=True, tol=None, verbose=0, warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'alpha': [1e-05, 0.0001, 0.001, 0.01, 0.1, 1, 10]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [28]:
clf = SGDClassifier(alpha=0.0001).fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.4f}".format(clf.score(test_document, Y_test)))

Training set score: 0.94
Test set score: 0.8859


C:\Users\sunilchawla\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


# <font color="red">hyperparameter tuning in LogisticRegression algorithm</font>

<font color="green">high value for the parameter C, Logis
ticRegression and LinearSVC try to fit the training set as best as possible, while with
low values of the parameter C, the models put more emphasis on finding a coefficient
vector (w) that is close to zero.</font>

In [19]:
param_grid = {'C': [0.01, 0.1, 1, 10,100]}
from sklearn.linear_model import LogisticRegression
grid_search = GridSearchCV(LogisticRegression(),param_grid, cv = 3)
grid_search.fit(train_document, Y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'C': [0.01, 0.1, 1, 10, 100]}, pre_dispatch='2*n_jobs',
       refit=True, return_train_score='warn', scoring=None, verbose=0)

In [24]:
clf = LogisticRegression(C = 1.0,penalty='l2',max_iter=1000).fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.4f}".format(clf.score(test_document, Y_test)))

Training set score: 0.93
Test set score: 0.8832


# <font color="red">hyperparameter tuning in TF-IDF</font>

In [82]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf_vect = TfidfVectorizer(ngram_range=(1, 3),min_df=2,max_df = 0.56)
train_document = tfidf_vect.fit_transform(X_train)
test_document = tfidf_vect.transform(X_test)

In [83]:
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression(C=12, penalty='l2').fit(train_document,Y_train)
print("Training set score: {:.2f}".format(clf.score(train_document, Y_train)))
print("Test set score: {:.4f}".format(clf.score(test_document, Y_test)))

Training set score: 1.00
Test set score: 0.9054


# <br><font color="blue">highest accuracy = 90.54% with hyperparameter tuning in TF-IDF and LogisticRegression(without data preprocessing) </font></br>